src/server/debug.rs

```diff
@@ -228,48 +228,69 @@ impl Debugger {
-
-        let store_id = self.get_store_id()?;
-        let peer_id = old_region
-            .get_peers()
-            .iter()
-            .find(|p| p.get_store_id() == store_id)
-            .map(|p| p.get_id())
-            .ok_or_else(|| {
-                Error::Other("RegionLocalState doesn't contains the peer itself".into())
-            })?;
-
-        let new_conf_ver = region.get_region_epoch().get_conf_ver();
-        let old_conf_ver = old_region.get_region_epoch().get_conf_ver();
-
-        // If the store is not in peers, or it's still in but its peer_id
-        // has changed, we know the peer is marked as tombstone success.
-        let scheduled = region
-            .get_peers()
-            .iter()
-            .find(|p| p.get_store_id() == store_id)
-            .map_or(true, |p| p.get_id() != peer_id);
-
-        if new_conf_ver > old_conf_ver && scheduled {
-            let wb = WriteBatch::new();
-            // Here we can keep the other metas as original.
-            box_try!(write_peer_state(db, &wb, &old_region, PeerState::Tombstone));
+        let wb = WriteBatch::new();
+        let handle = box_try!(get_cf_handle(db.as_ref(), CF_RAFT));
+        let mut ret = Vec::with_capacity(regions.len());
+
+        {
+            let fix_one_region = |region: Region| {
+                let id = region.get_id();
+                let key = keys::region_state_key(id);
+                let store_id = self.get_store_id()?;
+
+                let mut region_state = box_try!(db.get_msg_cf::<RegionLocalState>(CF_RAFT, &key))
+                    .expect("Can't find RegionLocalState");
+
+                let peer_id = region_state
+                    .get_region()
+                    .get_peers()
+                    .iter()
+                    .find(|p| p.get_store_id() == store_id)
+                    .map(|p| p.get_id())
+                    .ok_or_else(|| {
+                        Error::Other("RegionLocalState doesn't contains the peer itself".into())
+                    })?;
+
+                let old_conf_ver = region_state.get_region().get_region_epoch().get_conf_ver();
+                let new_conf_ver = region.get_region_epoch().get_conf_ver();
+                if new_conf_ver <= old_conf_ver {
+                    return Err(box_err!("invalid conf_ver"));
+                }
+
+                // If the store is not in peers, or it's still in but its peer_id
+                // has changed, we know the peer is marked as tombstone success.
+                let scheduled = region
+                    .get_peers()
+                    .iter()
+                    .find(|p| p.get_store_id() == store_id)
+                    .map_or(true, |p| p.get_id() != peer_id);
+                if !scheduled {
+                    return Err(box_err!("The peer is still in target peers"));
+                }
+
+                region_state.set_state(PeerState::Tombstone);
```

Comment:
Why not use `write_peer_state`?

Reply:
call `write_peer_state` is danguerous after we support `region merge`. I think that shouldn't be treated as a general function.

Reply:
How? `write_peer_state` should not be more dangerous than just put a state manually as the dangerous part are the same, which is overriding a meta key.

Reply:
`write_peer_state` will clean `merge_state` in `RegionLocalState`. So it will corrupt data if the region is merging.

Reply:
Does it really matter? If it matters, should it be checked before marking it as tombstone? I don't think it's reason not to use `write_peer_state`.